### PR TITLE
Slope shaping

### PR DESCRIPTION
**Pull request by trentgill**
Update: Complete & ready for testing then merge.

slopes library has been refactored:
- step function is broken into separate pieces according to assumptions about the sample block
- breakpoint blocks are handled recursively per-sample (the vector math was wonky before)
- a shaper function is applied at the end of the step function (can be block or persample)
- each slope is normalized to a [0,1] range and then remapped after shaping

TODO:
- [x] testing. initial reports suggest they sound great as envelopes!
- [x] add vectorized versions of the shapers (everything is done per-sample atm)
- [x] the shapers should be put into a library
- [x] add square shapes ('rise+wait' = `note()` in asllib, 'wait+rise' = `{to('here',t), to(d,0)}`
- [x] update docs for usage
- [x] add waveshape option to `lfo`, `ar` and `adsr` in asllib
- [x] `pulse` in asllib can be simplified with square-rising

completes first half of #68

### CODE REVIEW

**Comment by trentgill** [comment]
Usage updates are in the [PR on monome/docs](https://github.com/monome/docs/pull/210). Briefly, all slopes are now able to be *shaped* natively by the slopes C engine.

Options are:
* 'linear': previously the only option. ramp steadily to the destination value
* 'expo': gentle envelopes, and long feeling decays
* 'log': for aggressive or 'plucky' envelopes
* 'sine': nice for LFOs
* 'now': go immediately to the destination then wait for *time* seconds
* 'wait': wait for *time* seconds then go immediately to the destination
* 'over': similar to 'log' but overshoots before landing gently
* 'under': similar to 'expo' but first goes the wrong direction to 'gather momentum'
* 'rebound': simulates a bouncing ball

```lua
-- using the output.volts meta-action
-- eg: ramp to 2V over 200ms, with an exponential path
output[1].slew  = 0.2
output[1].shape = 'expo'
output[1].volts = 2.0

-- additional argument to the to() function
output[1].action = to( 3.0, 0.1, 'over' )

-- updated asllib functions
lfo( time, level, shape ) -- defaults to 'sine'
ar( attack, release, level, shape ) -- defaults to 'log'
adsr( attack, decay, sustain, release, shape ) -- defaults to 'linear'
```

**Comment by dndrks** [comment]
> @dndrks can you help doc this?

looks like trent beat me to it! I'll review the PR once this is live, tho!

**Comment by tehn** [comment]
perfect!

@trentgill this opens up some great new usage for crow, super excited to play it

**Comment by trentgill** [comment]
There's a binary [here](https://github.com/monome/crow/suites/340543728/artifacts/536081) thanks to @simonvanderveldt 's work on the github Actions :) Would love to have some folks try it before we go public, just to make sure the syntax & names are good, but also that the changes to asllib (lfo & ar shapes) are an improvement and not just showing off the capabilities.

**Comment by simonvanderveldt** [comment]
I'd like to help test, but will only have time either Friday or the coming weekend.

**Comment by dndrks** [comment]
@trentgill  -- just uploaded the .bin you posted and getting super wonky results with n2v in max and druid.

**experienced:**
`output[1].volts = n2v(3)` gives a super high note, executing again gives a super low note, high note, low note, etc.

**expected:**
`output[1].volts = n2v(3)` plays a note 3 semitones higher than base note

File: lua/asllib.lua [hunk 1]
```diff
@@ -15,18 +15,19 @@ function clamp( input, min, max )
 end
 
 function note( noteNum, duration )
-    return{ to( n2v(noteNum), 0 )
-          , to( 'here', duration  )
-          }
+    return to( n2v(noteNum), duration, 'now' )
```
Comment (dndrks) [inline lua/asllib.lua]: should this be `return{ to( n2v(noteNum), duration, 'now' ) }` ?

Comment (trentgill) [reply lua/asllib.lua]: This is fine- a standalone `to()` is a valid ASL by itself. You only need the table if you have more than one action. `to` is similar to the other constructs like `loop`, ie they are just functions that return ASLs `loop{ to(), to() }`.

Sorry I still haven't done the full write up of ASL :(

**Comment by trentgill** [comment]
Thanks for catching this bug @dndrks. The changes I made had broken `to()` calls with time=0 which is exactly what `output[n].volts = ...` is doing under the hood (when output.slew == 0). It's all working for me now. You can download a new bin by clicking `show all checks` immediately above this comment, and following through to 'build artifacts'.

**Comment by trentgill** [comment]
@simonvanderveldt 
> I'd like to help test, but will only have time either Friday or the coming weekend.
> 
> [edit] Quick question: How is "wait" different from the delay that was introduced in #249 ?

It's pretty orthogonal. This PR is about making fancy CV curves directly from the ASL syntax (by waveshaping the output samples). The `delay` function is about delaying execution of an arbitrary lua function. These shapes are all applied in the DSP loop (in C), and are only accessible through the list of names above.

For example. You can use the `lfo` asllib function to assign an LFO to an output. If you set the 'shape' to 'now', you get a square-wave that starts with the high-state. Conversely if you use 'wait', you get an LFO that starts in the low-state.

**Comment by dndrks** [comment]
@trentgill , I've tested about 6 hours in the last two days as I revamped the m4l devices to take advantage of new shapes. it all seems rock solid, really really into the expressivity these shapes open up! generally, everything feels really stable -- I only hung once after sending a metric shit ton of bad commands due to a patching mistake :)

**Comment by trentgill** [comment]
Great! Merging this now.

@dndrks re: naming. Why not just have a single canonical string for each? Under the hood, you only need as many chars as to make unambiguous (ie `e` for exponential, `lo` for logarithmic, `li` for linear), but that's obviously unreadable. The performance benefit of shorter names is likely dwarfed by the overhead of the lua call anyway. I think the list I posted in the examples comment above is probably good if accompanied by longer form docs?

**Pull request merged**